cpp/src/io/utilities/data_sink.cpp

```diff
@@ -49,36 +57,52 @@ class file_sink : public data_sink {
 
   size_t bytes_written() override { return _bytes_written; }
 
-  [[nodiscard]] bool supports_device_write() const override { return _cufile_out != nullptr; }
+  [[nodiscard]] bool supports_device_write() const override
+  {
+    return !_kvikio_file.closed() || _cufile_out != nullptr;
+  }
 
   [[nodiscard]] bool is_device_write_preferred(size_t size) const override
   {
-    return _cufile_out != nullptr && _cufile_out->is_cufile_io_preferred(size);
+    return !_kvikio_file.closed() ||
+           (_cufile_out != nullptr && _cufile_out->is_cufile_io_preferred(size));
   }
 
-  void device_write(void const* gpu_data, size_t size, rmm::cuda_stream_view stream) override
+  std::future<void> device_write_async(void const* gpu_data,
+                                       size_t size,
+                                       rmm::cuda_stream_view stream) override
   {
     if (!supports_device_write()) CUDF_FAIL("Device writes are not supported for this file.");
 
-    _cufile_out->write(gpu_data, _bytes_written, size);
+    size_t offset = _bytes_written;
     _bytes_written += size;
+
+    if (!_kvikio_file.closed()) {
+      // KvikIO's `pwrite()` returns a `std::future<size_t>` so we convert it
+      // to `std::future<void>`
+      return std::async(std::launch::deferred, [this, gpu_data, size, offset] {
+        _kvikio_file.pwrite(gpu_data, size, offset).get();
+      });
+    }
+    return _cufile_out->write_async(gpu_data, offset, size);
   }
 
-  std::future<void> device_write_async(void const* gpu_data,
-                                       size_t size,
-                                       rmm::cuda_stream_view stream) override
+  void device_write(void const* gpu_data, size_t size, rmm::cuda_stream_view stream) override
   {
     if (!supports_device_write()) CUDF_FAIL("Device writes are not supported for this file.");
 
-    auto result = _cufile_out->write_async(gpu_data, _bytes_written, size);
+    if (!_kvikio_file.closed()) {
+      return device_write_async(gpu_data, _bytes_written, stream).get();
+    }
+    _cufile_out->write(gpu_data, _bytes_written, size);
     _bytes_written += size;
```

Comment:
As I see, `_bytes_written` will not be updated if kvikio is used (due to the return at line 95).

Reply:
I would suggest copying lines 81-85 to line 95. Or, extract those lines into an inline function called by both places.

Reply:
Another solution (probably you would like) is to add comment to clarify before line 95.

Reply:
Actually, the two `_cufile_in->read` and `_cufile_out->write` also ends up calling their async counterparts: `cufile_input_impl::read_async()` and `cufile_output_impl::write_async`. I think we can simplify everything by calling the asyncs directly: https://github.com/rapidsai/cudf/pull/10593/commits/6f0b18498a72d71bd6dfc876c75e6bcc30faf5a3